### PR TITLE
Migrate to Typescript

### PR DESCRIPTION
**Pull request by prybalko**
`tsconfig.json` is taken from https://github.com/polkadot-js/dev/blob/master/packages/dev/config/tsconfig.json

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -50,7 +55,7 @@
   "browserslist": [
     "> 1%",
     "last 2 versions",
-    "not ie <= 8"
+    "not ie <= 11"
```
Comment (prybalko) [inline package.json:58]: `Crypto.getRandomValues(keys);` is not supported by older browsers
https://caniuse.com/getrandomvalues

Comment (AndreiEres) [reply package.json:58]: Do we really need IE? It's not supported even by Microsoft already.

File: src/main.ts [hunk 2]
```diff
@@ -1,8 +1,18 @@
 import Vue from "vue";
 
+declare module "vue/types/vue" {
```
Comment (AndreiEres) [inline src/main.ts:3]: Maybe better to move that to `typing.d.ts`

File: src/util/crypto.ts [hunk 3]
```diff
@@ -23,34 +37,34 @@ const HexEncodeArray = [
   "f"
 ];
 
-function strToUint8Array(str) {
-  return new TextEncoder("utf-8").encode(str);
+function strToUint8Array(str: string): Uint8Array {
+  return new TextEncoder().encode(str);
```
Comment (AndreiEres) [inline src/util/crypto.ts:41]: It's correct that we removed `"utf-8"` here?

File: src/util/crypto.ts [hunk 4]
```diff
@@ -122,38 +151,39 @@ function reconstruct(shardObjects, passphrase) {
...
   }
 
   switch (versions[0]) {
     case 0:
-      var shardData = shardObjects.map(shard => shard.data);
-      var encryptedSecret = SECRETS.combine(shardData);
-      var secret = dehexify(encryptedSecret);
-      var nonce = dehexify(nonces[0]);
-      var salt = hashString(titles[0]);
+      var shardData = shardObjects.map(shard => shard.data),
```
Comment (AndreiEres) [inline src/util/crypto.ts:171]: Usually we start each declaration with keyword. But it's on your linter. 
And is `var` ok?

File: src/views/Combine.vue [hunk 5]
```diff
@@ -53,69 +53,82 @@
...
+};
+
+export default Vue.extend({
   name: "Combine",
-  data: function() {
+  data(): CombineData {
     return {
       title: "",
       nonce: "",
-      shards: new Set(),
```
Comment (AndreiEres) [inline src/views/Combine.vue:67]: It's correct?

**Review by AndreiEres — COMMENTED** [review]
lgtm, but i'm not vue expert. 
Also I thought you wanted cut off vue. So looks strange to rewrite it to ts before cutting.

File: src/util/crypto.ts (hunk 3 above)
Comment (prybalko) [reply src/util/crypto.ts:41]: Yes, I think so. `TextEncoder` constructor doesn't take any arguments and always uses `utf-8`
https://developer.mozilla.org/en-US/docs/Web/API/TextEncoder

File: src/views/Combine.vue (hunk 5 above)
Comment (prybalko) [reply src/views/Combine.vue:67]: Yes, I slightly changed the logic: now `qrCodes` is a Set and `shards` is an array.

File: package.json (hunk 1 above)
Comment (prybalko) [reply package.json:58]: I don't know.
cc: @kirushik @vas3k

File: src/util/crypto.ts (hunk 4 above)
Comment (prybalko) [reply src/util/crypto.ts:171]: > Usually we start each declaration with keyword

Yeah, that's the matter of preference. Staring each declaration with keyword seems excessively verbose. But if that's the common practice, I'll change it back.

P.S. Currently linter has no rules about it.

File: src/main.ts (hunk 2 above)
Comment (prybalko) [reply src/main.ts:3]: Thanks a good idea, thanks

**Review by vas3k — APPROVED** [review]

File: package.json (hunk 1 above)
Comment (kirushik) [reply package.json:58]: This was broken anyway — so I'm OK to drop IE8..10 support.
In general, I'd like BananaSplit to be as conservative in browser support as possible — but only as long as doing so won't introduce any potential attack vectors: say, adding an extra "shim" dependency opens us up for a supply chain attack, and is worse than browser support degradation. (But the actual _feature_ which requires this browser API should be considered very carefully in such case.)

It should all be a reflection of our big design goal to guarantee Banana Split extremely long "shelf life", ideally 10–20 years for the released htmls without any modifications being needed.

**Review by kirushik — APPROVED** [review]
I'm OK with this PR; originally I tried to avoid TypeScript to reduce the amount of tooling (and associated trust required) to build the project — but probably additional typing-based assurance of correctness probably wins, especially for the long-term supported project.

LGTM

**Pull request merged**